### ISSUE TITLE
NET-1589 Fix FP S2589: Assignment for captures are ignored

### ISSUE THREAD

**Issue by martin-strecker-sonarsource** [SonarSource/sonar-dotnet#9204]
Originally reported at https://community.sonarsource.com/t/false-positive-for-s2589/114185

### Description

Any mutation of a captured variable is ignored. This applies to captures by delegates and by local functions. See also https://github.com/SonarSource/sonar-dotnet/issues/8885

### Repro steps

```cs
public bool ForEachTest(List<string> licenseData)
{
    var found = false;
    licenseData.ForEach(license => found = true); // Assignment in "ForEach"
    if (!found) // Noncompliant FP
    {
        Console.WriteLine("No License for artifact type");
    }
    return found;
}

public bool SelectTest(List<string> licenseData)
{
    var found = false;
    licenseData.Select(license => found = true).Any(); // Assignment in "Select"
    if (!found) // Noncompliant FP
    {
        Console.WriteLine("No License for artifact type");
    }
    return found;
}

public bool ActionTest()
{
    var found = false;
    Action assign = () => found = true; // Assignment in some delegate
    assign();
    if (!found) // Noncompliant FP
    {
        Console.WriteLine("No License for artifact type");
    }
    return found;
}

public bool LocalFunctionTest()
{
    var found = false;
    Assign();
    if (!found) // Noncompliant FP
    {
        Console.WriteLine("No License for artifact type");
    }
    return found;

    void Assign() => found = true; // Assignment in local function
}
```
### Expected behavior

Do not raise on mutated captured variables or learn inside the delegate/local function body.

### Actual behavior

FP for S2589

### Known workarounds

Do not mutate capture variables.

**Comment by martin-strecker-sonarsource** [comment]
## Possible solutions

### Invocation unrolling

See https://trello.com/c/dELDhBru

### Ignoring captured and mutated symbols after their capturing

* Handle [FlowAnonymousFunctionOIperation](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.flowanalysis.iflowanonymousfunctionoperation?view=roslyn-dotnet-4.7.0) and get the CFG (including nested)
* Find assignments and mutations of captured symbols (symbols we know about in ProgramState)
* Then do either
  * Unlearn all constraints (see case below where this is problematic)
  * Remember these symbols as being problematic. The engine stops tracking those. Single rules may need to take care of these "untrustworthy" symbols before reporting because they may use states from before we learned about the capture being problematic.

### Ignoring captured and mutated symbols completely

* Like before but as a pre-processing step before the engine starts for the outer method


```cs
public bool SelectTest(List<string> licenseData)
{
    bool found;
    var qry = licenseData.Select(license => found = true)
    found = false;
    qry.Any();  // This evaluates the lambda and maybe sets true. This is hard to predict because the lambda cfg is visited before the found = false
    if (!found) // Noncompliant FP
    {
        Console.WriteLine("No License for artifact type");
    }
    return found;
}
```

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1589](https://sonarsource.atlassian.net/browse/NET-1589)

[NET-1589]: https://sonarsource.atlassian.net/browse/NET-1589?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
